storage/txcache/evictionStrategy.go

```diff
@@ -0,0 +1,168 @@
+package txcache
+
+import (
+	"sync"
+
+	"github.com/ElrondNetwork/elrond-go/core"
+	"github.com/ElrondNetwork/elrond-go/data/transaction"
+	"github.com/ElrondNetwork/elrond-go/logger"
+)
+
+var log = logger.GetOrCreate("txcache/eviction")
+
+// EvictionStrategyConfig is a cache eviction model
+type EvictionStrategyConfig struct {
+	CountThreshold                 int
+	CountJustAFewSenders           int
+	NoOldestSendersToEvict         int
+	ALotOfTransactionsForASender   int
+	NoTxsToEvictForASenderWithALot int
+}
+
+// EvictionStrategy is a cache eviction model
+type EvictionStrategy struct {
+	Cache  *TxCache
+	Config EvictionStrategyConfig
+	mutex  sync.Mutex
+}
+
+// NewEvictionStrategy creates a new EvictionModel
+func NewEvictionStrategy(cache *TxCache, config EvictionStrategyConfig) *EvictionStrategy {
+	model := &EvictionStrategy{
+		Cache:  cache,
+		Config: config,
+	}
+
+	return model
+}
+
+// DoEviction does cache eviction
+func (model *EvictionStrategy) DoEviction(incomingTx *transaction.Transaction) {
+	if !model.areThereTooManyTxs() {
+		return
+	}
+
+	// We do not allow more evictions to start concurrently
+	model.mutex.Lock()
+
+	// First pass
+	// Senders capacity is close to be reached first (before txs capacity) when there are a lot of senders with little or one transaction
+	if model.areThereTooManySenders() {
+		countTxs, countSenders := model.EvictOldestSenders()
+		log.Debug("DoEviction, 1st pass:", "countTxs", countTxs, "countSenders", countSenders)
+	}
+
+	// Second pass
+	if model.areThereTooManyTxs() {
+		countTxs, countSenders := model.EvictHighNonceTransactions()
+		log.Debug("DoEviction, 2nd pass:", "countTxs", countTxs, "countSenders", countSenders)
+	}
+
+	// Third pass
+	if model.areThereTooManyTxs() && !model.areThereJustAFewSenders() {
+		steps, countTxs, countSenders := model.EvictSendersWhileTooManyTxs()
+		log.Debug("DoEviction, 3rd pass:", "steps", steps, "countTxs", countTxs, "countSenders", countSenders)
+	}
+
+	model.mutex.Unlock()
+}
+
+func (model *EvictionStrategy) areThereTooManySenders() bool {
+	noSenders := model.Cache.txListBySender.Counter.Get()
```

Comment:
this is too long. too many structs inside structs.

Reply:
I plead guilty for breaking the Law of Demeter.

Reply:
Fixed for `areThereTooManySenders`, `areThereJustAFewSenders`, `areThereTooManyTxs`.